packages/cli/src/commands/governance/upvote.ts

```diff
@@ -29,6 +30,43 @@ export default class Upvote extends BaseCommand {
       .runChecks()
 
     const account = await (await this.kit.contracts.getAccounts()).voteSignerToAccount(signer)
-    await displaySendTx('upvoteTx', await governance.upvote(id, account), {}, 'ProposalUpvoted')
+
+    const consideredProposals = await this.dequeueAllPossibleProposals(governance as any)
+
+    if (!consideredProposals.some((k) => k.id === id)) {
+      await displaySendTx('upvoteTx', await governance.upvote(id, account), {}, 'ProposalUpvoted')
+    }
+  }
+
+  /**
+   * Dequeues all possible proposals, returns the ones that were considered to be dequeued.
+   */
+  async dequeueAllPossibleProposals(governance: GovernanceWrapper) {
+    const concurrentProposals = (await governance.concurrentProposals()).toNumber()
+    const queue = await governance.getQueue()
+    const originalLastDequeue = await governance.lastDequeue()
+
+    let consideredProposals
+
+    for (let index = 0; index < queue.length / concurrentProposals + 1; index++) {
+      consideredProposals = (
+        await Promise.all(
+          queue
+            .slice(0, concurrentProposals)
+            .map((p) => p.proposalID)
+            .map(async (id) => {
+              const expired = await governance.isQueuedProposalExpired(id)
+              return { id: id.toString(), expired }
+            })
+        )
+      ).filter((k) => k.expired === false)
+
+      await displaySendTx('dequeue', governance.dequeueProposalsIfReady(), {})
```

Comment:
has this to be inside of the loop? Because on-chain one tx should be enough to dequeue them all, this function should only check if there's on that needs to be dequeued.

This is specially hurtful if the user is using a ledger.

Reply:
This cycle is there only for safety purposes (in 99.9% of times it will not be run more than once). 

The reason behind this cycle is as follows:

1. Someone will DDoS our proposals (eg 100+)
2. All these proposals will expire 
3. DequeueProposalsIfReady will always try to dequeue first n proposals but if all of them will be expired it will just remove them from queue.
4. Since none of the proposals were actually dequeued, next call will allow to dequeue again
5. Upvote function will try to dequeue again and possibly it will hit the proposal and bug that we have

Since this case is extremely unlikely we didn't consider it before at all. Now it would cause some inconvenience but it will behave deterministically. 

In general I believe it is extremely unlikely that this use case should happen, but it felt like a little work to cover it just in case.

Reply:
I added this explanation to method description